crates/oxc_parser/src/lexer/number.rs

```diff
@@ -83,72 +82,121 @@ fn parse_binary(s: &str) -> f64 {
     result as f64
 }
 
+// ==================================== OCTAL ====================================
+
+/// b'0' is 0x30 and b'7' is 0x37.
+///
+/// So we can convert from any octal digit to its value with `c & 7`.
+/// This is produces more compact assembly than `c - b'0'`.
+///
+/// <https://godbolt.org/z/9rYTsMoMM>
+const fn octal_byte_to_value(c: u8) -> u8 {
+    debug_assert!(c >= b'0' && c <= b'7');
+    c & 7
+}
 #[allow(clippy::cast_precision_loss, clippy::cast_lossless)]
 fn parse_octal(s: &str) -> f64 {
-    // b'0' is 0x30 and b'7' is 0x37.
-    // So we can convert from any octal digit to its value with `c & 7`.
-    // This is produces more compact assembly than `c - b'0'`.
-    // https://godbolt.org/z/9rYTsMoMM
-    const fn byte_to_value(c: u8) -> u8 {
-        debug_assert!(c >= b'0' && c <= b'7');
-        c & 7
-    }
-    #[cfg(test)]
-    {
-        static_assertions::const_assert_eq!(byte_to_value(b'0'), 0);
-        static_assertions::const_assert_eq!(byte_to_value(b'7'), 7);
+    const MAX_FAST_OCTAL_LEN: usize = 21;
+
+    debug_assert!(!s.is_empty());
+    debug_assert!(!s.starts_with("0o") && !s.starts_with("0O"));
+
+    let without_leading_zeros = s.trim_start_matches('0');
```

Comment:
I think this logic can be simplified to minimize the the number of operations before getting into the fast path (which will be almost all cases). Just this would do it:

```rs
if s.len() > MAX_FAST_OCTAL_LEN {
  parse_octal_slow(s)
} else {
  // Fast path code here
}
```

I know it was me who originally suggested trimming off leading zeros, but with the simpler (better) version of `parse_octal_slow` that you've used, it's not necessary. `trim_start_matches` is one of the annoying stdlib functions which processes text as `char`s, so it's quite slow. If you wanted to retain the removal of leading zeros, you could implement it as loop over bytes.

Reply:
Done